pkg/ddevapp/commands.go

```diff
@@ -1,47 +1,83 @@
+	// Remove legacy directory
+	legacyCommandDir := app.GetConfigPath(".global_commands")
+	if fileutil.IsDirectory(legacyCommandDir) {
+		err := fileutil.PurgeDirectory(legacyCommandDir)
+		if err != nil {
+			util.Warning("Unable to remove files from legacy directory %s: %v", legacyCommandDir, err)
+			return nil
+		}
+		err = os.Remove(legacyCommandDir)
+		if err != nil {
+			util.Warning("Unable to remove legacy directory %s: %v", legacyCommandDir, err)
+		}
+	}
 
 	sourceGlobalCommandPath := filepath.Join(globalconfig.GetGlobalDdevDir(), "commands")
 	err := os.MkdirAll(sourceGlobalCommandPath, 0755)
 	if err != nil {
 		return nil
 	}
 
-	projectCommandPath := app.GetConfigPath("commands")
-	// Make sure our target global command directory is empty
-	copiedGlobalCommandPath := app.GetConfigPath(".global_commands")
-	err = os.MkdirAll(copiedGlobalCommandPath, 0755)
+	commandDirs, err := fileutil.ListFilesInDirFullPath(sourceGlobalCommandPath)
 	if err != nil {
-		util.Error("Unable to create directory %s: %v", copiedGlobalCommandPath, err)
-		return nil
+		return err
 	}
 
-	// Make sure it's empty
-	err = fileutil.PurgeDirectory(copiedGlobalCommandPath)
-	if err != nil {
-		util.Error("Unable to remove %s: %v", copiedGlobalCommandPath, err)
-		return nil
-	}
+	for _, serviceDirOnHost := range commandDirs {
+		// If the item isn't a directory, skip it.
+		if !fileutil.IsDirectory(serviceDirOnHost) {
+			continue
+		}
 
-	err = copy2.Copy(sourceGlobalCommandPath, copiedGlobalCommandPath)
-	if err != nil {
-		return err
-	}
+		service := filepath.Base(serviceDirOnHost)
+		containerName := GetContainerName(app, service)
+		commandDirInService := path.Join("/mnt/ddev-global-cache/commands/", app.Name, service)
 
-	if !fileutil.FileExists(projectCommandPath) || !fileutil.IsDirectory(projectCommandPath) {
-		return nil
+		// Skip if a container for that service doesn't exist
+		container, err := dockerutil.FindContainerByName(containerName)
+		if err != nil || container == nil {
+			continue
+		}
+
+		// Remove contents of the directory, if the directory exists and has some contents
+		uid, _, _ := util.GetContainerUIDGid()
+		_, stderr, err := dockerutil.Exec(container.ID, "[[ $(ls -A "+commandDirInService+") ]] && rm "+path.Join(commandDirInService, "*")+" || true", uid)
+		if err != nil {
+			return fmt.Errorf("unable to rm %s inside %s: %v (stderr=%s)", commandDirInService, containerName, err, stderr)
+		}
+
+		commandFiles, err := fileutil.ListFilesInDir(serviceDirOnHost)
+		if err != nil {
+			return err
+		}
+		// Any command we find will want to be executable on Linux
+		for _, commandName := range commandFiles {
+			onHostFullPath := filepath.Join(serviceDirOnHost, commandName)
+			_ = os.Chmod(onHostFullPath, 0755)
+		}
+		// Copy commands into container (this will create the directory if it's not there already)
+		err = dockerutil.CopyIntoContainer(serviceDirOnHost, containerName, commandDirInService, "")
```

Comment:
I would rather see one call here rather than multiple calls. That's an argument in favor of just copying everything. We already do too many `CopyIntoContainer` calls. Maybe it doesn't matter so much.

Reply:
Yeah I probably will just copy everything - I had it this way because originally it wasn't copying into a shared volume, so with that implementation it made sense for each container to only have it's own commands.

Reply:
Copying everything _except_ host now. Copying host could lead to confusion at best, as those will only ever be run in place in the `.global_commands` dir.
All service commands are copied regardless of whether there's a container for that service in the project (mostly 'cause that was easier)

Reply:
Does still need a separate call to delete the old commands dir though, so we don't have commands in the volume that don't exist in `~/.ddev/commands` anymore.